// mobile-mcp/dist/mcp_server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.startMCPServer = startMCPServer;
const mcp_js_1 = require("@modelcontextprotocol/sdk/server/mcp.js");
const stdio_js_1 = require("@modelcontextprotocol/sdk/server/stdio.js");
const zod_1 = require("zod");
const mobile_use_1 = require("mobile-use");
const adb = new mobile_use_1.ADBClient();
// Create an MCP server
const server = new mcp_js_1.McpServer({
    name: "Mobile MCP",
    version: "0.0.7",
}, {
    instructions: `Before doing anything, you need to initialize the mobile by using mobile_init. 
<important_notes>
Use mobile_dump_ui for navigating and interacting with the apps. Use mobile_screenshot as fallback to understand the context.
</important_notes>
    `,
});
// Initialize ADB client
// ADB Tools
server.tool("mobile_dump_ui", "Get a detailed hierarchy of all interactive UI elements on the current screen, including their positions, text, and attributes. Use this when you want to navigate or take actions.", {}, async () => {
    const ui = await adb.dumpUI();
    return { content: [{ type: "text", text: ui }] };
});
server.tool("mobile_tap", "Simulate a tap gesture at specific x,y coordinates on the screen to interact with UI elements", { x: zod_1.z.number(), y: zod_1.z.number() }, async ({ x, y }) => {
    await adb.tap({ x, y });
    return { content: [{ type: "text", text: `Tapped at (${x}, ${y})` }] };
});
server.tool("mobile_swipe", "Perform a swipe gesture from one coordinate to another, optionally specifying duration in milliseconds", {
    startX: zod_1.z.number(),
    startY: zod_1.z.number(),
    endX: zod_1.z.number(),
    endY: zod_1.z.number(),
    duration: zod_1.z.number().optional(),
}, async ({ startX, startY, endX, endY, duration }) => {
    await adb.swipe({ x: startX, y: startY }, { x: endX, y: endY }, duration);
    return {
        content: [
            {
                type: "text",
                text: `Swiped from (${startX}, ${startY}) to (${endX}, ${endY})`,
            },
        ],
    };
});
server.tool("mobile_type", "Input text into the currently focused text field or input box", { text: zod_1.z.string() }, async ({ text }) => {
    await adb.type(text);
    return { content: [{ type: "text", text: `Typed: ${text}` }] };
});
server.tool("mobile_key_press", "Simulate pressing a specific key or button (e.g., 'enter', 'back', 'home')", { key: zod_1.z.string() }, async ({ key }) => {
    await adb.keyPress(key);
    return { content: [{ type: "text", text: `Pressed key: ${key}` }] };
});
server.tool("mobile_screenshot", "Capture the current screen state as an image, useful for visual verification or when UI hierarchy is insufficient.", {}, async () => {
    const screenshot = await adb.screenshot();
    const size = await adb.screenSize();
    return {
        content: [
            {
                type: "text",
                text: `Screenshot captured at width: ${size.width}px and height: ${size.height}px`,
            },
            {
                data: screenshot.toString("base64"),
                mimeType: "image/png",
                type: "image",
            },
        ],
    };
});
server.tool("mobile_list_packages", "List all installed packages on the device, optionally filtered by a search term", { filter: zod_1.z.string().optional() }, async ({ filter }) => {
    const packages = await adb.listPackages(filter);
    return { content: [{ type: "text", text: JSON.stringify(packages) }] };
});
server.tool("mobile_open_app", "Open an application using its package name", { packageName: zod_1.z.string() }, async ({ packageName }) => {
    await adb.openApp(packageName);
    return { content: [{ type: "text", text: `Opened app: ${packageName}` }] };
});
server.tool("mobile_init", "Use this command to initalize mobile device.", {}, async () => {
    try {
        await adb.init();
        return {
            content: [
                {
                    type: "text",
                    text: `Successfully initialized mobile devices.`,
                },
            ],
        };
    }
    catch (error) {
        return {
            content: [
                {
                    type: "text",
                    text: `Couldn't initalize the device.\n${error.message}`,
                },
            ],
        };
    }
});
async function startMCPServer() {
    try {
        const transport = new stdio_js_1.StdioServerTransport();
        await server.connect(transport);
        console.error("Mobile use mcp server is running.");
    }
    catch (error) {
        console.error("Error starting MCP server:", error);
    }
}
